atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.cassandra.thrift.CqlResult;
+import org.apache.cassandra.thrift.InvalidRequestException;
+import org.apache.cassandra.thrift.KeyRange;
+import org.apache.cassandra.thrift.KeySlice;
+import org.apache.cassandra.thrift.Mutation;
+import org.apache.cassandra.thrift.SchemaDisagreementException;
+import org.apache.cassandra.thrift.SlicePredicate;
+import org.apache.cassandra.thrift.TimedOutException;
+import org.apache.cassandra.thrift.UnavailableException;
+import org.apache.thrift.TException;
+
+import com.palantir.atlasdb.qos.AtlasDbQosClient;
+import com.palantir.common.base.Throwables;
+import com.palantir.processors.AutoDelegate;
+
+/**
+ * Wrapper for Cassandra.Client.
+ */
+@AutoDelegate(typeToExtend = Cassandra.Client.class)
+@SuppressWarnings({"checkstyle:all", "DuplicateThrows"}) // :'(
+public class CassandraClient extends AutoDelegate_Client {
+    private final Cassandra.Client delegate;
+    private final AtlasDbQosClient qosClient;
+
+    public CassandraClient(Cassandra.Client delegate, AtlasDbQosClient qosClient) {
+        super(delegate.getInputProtocol());
+        this.delegate = delegate;
+        this.qosClient = qosClient;
+    }
+
+    @Override
+    public Cassandra.Client delegate() {
+        return delegate;
+    }
+
+    @Override
+    public Map<ByteBuffer, List<ColumnOrSuperColumn>> multiget_slice(List<ByteBuffer> keys, ColumnParent column_parent,
+            SlicePredicate predicate, ConsistencyLevel consistency_level)
+            throws InvalidRequestException, UnavailableException, TimedOutException, TException {
+        return checkLimitAndCall(() -> super.multiget_slice(keys, column_parent, predicate, consistency_level));
+    }
+
+    @Override
+    public void batch_mutate(Map<ByteBuffer, Map<String, List<Mutation>>> mutation_map,
+            ConsistencyLevel consistency_level)
+            throws InvalidRequestException, UnavailableException, TimedOutException, TException {
+        checkLimitAndCall(() -> {
+            super.batch_mutate(mutation_map, consistency_level);
+            return null;
+        });
+    }
+
+    @Override
+    public CqlResult execute_cql3_query(ByteBuffer query, Compression compression, ConsistencyLevel consistency)
+            throws InvalidRequestException, UnavailableException, TimedOutException, SchemaDisagreementException,
+            TException {
+        return checkLimitAndCall(() -> super.execute_cql3_query(query, compression, consistency));
+    }
+
+    @Override
+    public List<KeySlice> get_range_slices(ColumnParent column_parent, SlicePredicate predicate, KeyRange range,
+            ConsistencyLevel consistency_level)
+            throws InvalidRequestException, UnavailableException, TimedOutException, TException {
+        return checkLimitAndCall(() -> super.get_range_slices(column_parent, predicate, range, consistency_level));
+    }
+
+    private <T> T checkLimitAndCall(Callable<T> callable) {
```

Comment:
what's the purpose of accepting a callable here? why not just `checkLimit()` before each call?

Reply:
Got rid of this method.